### PR TITLE
Java: Tests for Automodel Extraction Queries

### PR DESCRIPTION
**Pull request by kaeluka**
This PR adds some tests for the automodel extraction queries. `@adityasharad` will review this, but other reviewers are always welcome.

 - The tests come with hand-written Java code, so that they can also serve as documentation of intent (see the comments in the java files).
 - This does only add tests, not change the queries themselves (edit: an exception: fixed a badly named query).
 - Taus suggested, in the (backref'd) issue to add some models just for testing. I've found this to not be necessary, as we can easily test modeled classes by 're-implementing' them. Eg., there's a class `java.io.File` implemented in the test suite. I find this much easier than the suggested implementation. I was hoping this would work, and I'm glad it does.

How to review/Questions for the reviewers:

 - We'll add more test cases to these suites as we maintain the queries, but are there important features you'd like to see tested straight away?
 - Do you have out of the box ideas to improve this?
 - The intention is to set up a *small* test-suite that'll make it easier to add tests of new and existing behaviour in the future.
   The reason why this only tests *some* of our queries (only the extraction queries; the queries share most of the implementations behind the scenes) is to manage the maintenance overhead of the test suites. LMK if you disagree with that and would like to see more test suites.

### CODE REVIEW

**Comment by kaeluka** [comment]
After merging https://github.com/github/codeql/pull/13747, this needs minor updates. This should not block the review imho 👍

**Comment by kaeluka** [comment]
After merging https://github.com/github/codeql/pull/13747, this needed minor updates 👍

File: java/ql/test/query-tests/Telemetry/AutomodelApplicationModeExtractCandidates/Test.java
Comment (adityasharad) [inline java/ql/test/query-tests/Telemetry/AutomodelApplicationModeExtractCandidates/Test.java:11]: High-level comment: Does it make sense to have one test directory, with a source file, and then 3 qlref files (pointing to each of the 3 queries)? Then for the same test source file we will see what gets considered a positive example, negative example, or candidate. This could save you some source code and library stub duplication across the test cases, while also allowing us to see in one place what the different candidates are.

File: java/ql/test/query-tests/Telemetry/AutomodelFrameworkModeExtractCandidates/java/nio/file/Files.java
Comment (adityasharad) [inline java/ql/test/query-tests/Telemetry/AutomodelFrameworkModeExtractCandidates/java/nio/file/Files.java:1]: There is a `stubs` directory generally used for such stub library definitions. Could this be placed there? I see no existing examples of stubs for JDK classes so worth checking with the Java team how they usually handle these - it may be that we assume the JDK is already a dependency and so is extracted.

**Review by adityasharad — COMMENTED** [review]
Glad to see tests! Some high level comments on structure before I get into checking the test cases themselves.

File: java/ql/test/query-tests/Telemetry/AutomodelApplicationModeExtractCandidates/Test.java
Comment (kaeluka) [reply java/ql/test/query-tests/Telemetry/AutomodelApplicationModeExtractCandidates/Test.java]: I think that's a great question/suggestion. When starting, my gut feeling was that it's a good idea to follow established conventions in the pre-existing tests (one suite per .ql file), and that we could get better readability from keeping things apart — but it turns out that the examples I used ended up being pretty similar. I've now pushed two commits to merge the three application-mode tests into one and the three framework-mode tests into one 👌

File: java/ql/test/query-tests/Telemetry/AutomodelFrameworkModeExtractCandidates/java/nio/file/Files.java
Comment (kaeluka) [reply java/ql/test/query-tests/Telemetry/AutomodelFrameworkModeExtractCandidates/java/nio/file/Files.java]: after a bit of back and forth, I got this to work :) [2a00bec](https://github.com/github/codeql/pull/13788/commits/2a00becc3c29de3d447f1d6fabdd24d837bd9a85)

**Comment by kaeluka** [comment]
following @adityasharad's comments, this has been restructured significantly. Before merging this, I'd like to squash these commits a bit.

File: java/ql/test/query-tests/Telemetry/AutomodelFrameworkModeExtractCandidates/java/nio/file/Files.java
Comment (kaeluka) [reply java/ql/test/query-tests/Telemetry/AutomodelFrameworkModeExtractCandidates/java/nio/file/Files.java]: @atorralba or @jcogs33 - could one of you confirm that we're using the stubs directory correctly? Or would you prefer we revert commit [2a00bec](https://github.com/github/codeql/pull/13788/commits/2a00becc3c29de3d447f1d6fabdd24d837bd9a85)?

Comment (jcogs33) [reply java/ql/test/query-tests/Telemetry/AutomodelFrameworkModeExtractCandidates/java/nio/file/Files.java]: JDK stubs shouldn't be needed normally. I _think_ what you're doing is similar to the test case I created on [this PR](https://github.com/github/codeql/pull/12680#issuecomment-1496665864). :smile: 
IMO, these stubs probably should not live in the `stubs` directory since they're essentially being used as source code, and they're specific to one test case instead of being generic stubs (but I'll defer to @atorralba's opinion on placement).

Comment (atorralba) [reply java/ql/test/query-tests/Telemetry/AutomodelFrameworkModeExtractCandidates/java/nio/file/Files.java]: I definitely think these shouldn't be in the `stubs` directory to not add confusion about the necessity of JDK stubs. In this case, since these are a few calls only, I'd prefer a source file in the test directory (say, `A.java`) that makes the specific calls you want to test the methods of. That way you ensure that they're extracted, and then your CodeQL code should be able to find the method definitions.

Comment (kaeluka) [reply java/ql/test/query-tests/Telemetry/AutomodelFrameworkModeExtractCandidates/java/nio/file/Files.java]: OK, I sort of expected this answer :) I have removed the last commit that moved those tests to the stubs directory. Resolving this conversation :)

File: java/ql/test/query-tests/Telemetry/AutomodelFrameworkModeExtraction/com/github/codeql/test/PublicClass.java
Comment (adityasharad) [inline java/ql/test/query-tests/Telemetry/AutomodelFrameworkModeExtraction/com/github/codeql/test/PublicClass.java:13]: Add an example with default visibility too (no visibility modifier).

**Review by adityasharad — COMMENTED** [review]
Good stuff. A few minor comments.

File: java/ql/test/query-tests/Telemetry/AutomodelFrameworkModeExtraction/com/github/codeql/test/NonPublicClass.java
Comment (kaeluka) [reply java/ql/test/query-tests/Telemetry/AutomodelFrameworkModeExtraction/com/github/codeql/test/NonPublicClass.java:8]: making this class would lead to the `String here` parameter being selected as a candidate. Why do you think this means the example code should be tweaked?

Comment (kaeluka) [reply java/ql/test/query-tests/Telemetry/AutomodelFrameworkModeExtraction/com/github/codeql/test/NonPublicClass.java:8]: Making this class `public` would lead to the `String here` parameter being selected as a candidate. Why do you think this means the example code should be tweaked?

Comment (kaeluka) [reply java/ql/test/query-tests/Telemetry/AutomodelFrameworkModeExtraction/com/github/codeql/test/NonPublicClass.java:8]: Further implementation detail: the reason is that the class is not considered externally visible by the `ModelApi` class that `@jcogs33` has built. We use this class to not select candidates that won't be callable from the outside.

**Comment by kaeluka** [comment]
@adityasharad, do you think this can be merged? If so, do you think I should squash these commits a bit, or would you merge as-is?

**Review by adityasharad — APPROVED** [review]
Looks good. I also verified that the new tests did actually run in the Java Language Tests run. Commits are clear enough I think, no need to squash.

**Pull request merged**
